Readers/src/EventReader.cc

```diff
@@ -482,6 +483,7 @@ EventReader::read() const
   {
     event_.muons_ = muonReader_->read();
     event_.muon_ptrs_ = convert_to_ptrs(event_.muons_);
+    setFilterbits(event_.muon_ptrs_, *(event_.triggerInfo_));
```

Comment:
Is the `TriggerInfo` object only used to set the filter bits here and never used again? If so, then I'd remove the `triggerInfo_` member from the `Event` class definition and get rid of all those dereferencing and addressing operations that are done to this object in multiple places.

Reply:
I would vote that we replace *(event_.triggerInfo_) by event_.triggerInfo() .

I would also vote to move the function
  setFilterbits(const std::vector<const T*>& particles, const TriggerInfo& triggerInfo) const
into its own setFilterbits.h file (I see no reason why this function needs to be defined in the EventReader.h file, as the setFilterbits function does not seem to access any information besides the lepton collection and TriggerInfo object passed as function arguments, i.e. the setFilterbits function can me made a "standalone" global function rather than a member function of the EventReader class).

Reply:
Hi, the best solution to that problem seems to me to:
- add a function void RecoLepton::set_filterBits(Int_t filterBits)
- add a function void RecoHadTau::set_filterBits(Int_t filterBits)
adding these functions will not require adding a friend declaration. 
We use similar functions for RecoLepton::set_p4() and RecoLepton::set_mvaRawTTH_cut() already

Reply:
I think this is a better approach indeed, but I also don't see a good reason to keep `TriggerInfo` as part of the `Event` class. We only need this object to set the HLT filter bits on the leptons and taus (which are then read by `TriggerInfoWriter`). After `EventReader::read()` is done, we can dispose of `TriggerInfo` (right?). Even the debugging / logging stuff could be moved into the `set_filterBits` member functions if needed.

Reply:
Do I implement setFilterbit as member function of RecoLepton and RecoHadTau? I don't know any specific reason of doing this

Reply:
The reason why we can call `set_p4()` and `set_mvaRawTTH_cut()` on `RecoLepton` objects is because these calls are done before they're "converted" into const pointers. If we want to avoid running `setFilterBits` on the actual objects instead of their pointers, then the only way to get rid of the `friend class` statements and `mutable` declarations is by moving these calls into the `read` functions of `RecoElectronReader`, `RecoMuonReader` and `RecoHadTauReader`. Whatever the case may be, there should still be one global function.

I can prepare a PR after this one's merged, but I don't think it's strictly necessary. I'll approve the current changes, but leave the merging up to Christian.